File: src/main/java/com/vandendaelen/nicephore/thread/InitThread.java
package com.vandendaelen.nicephore.thread;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;
import com.vandendaelen.nicephore.utils.Reference;
import com.vandendaelen.nicephore.utils.Util;
import net.minecraft.client.Minecraft;
import org.apache.commons.compress.utils.FileNameUtils;
import org.apache.commons.io.FileUtils;
import org.rauschig.jarchivelib.Archiver;
import org.rauschig.jarchivelib.ArchiverFactory;

import java.io.*;
import java.lang.reflect.Type;
import java.net.URL;
import java.net.URLConnection;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Collection;
import java.util.Optional;

public final class InitThread extends Thread {

    private boolean optimiseConfig;

    private static final File
            DESTINATION = new File(Minecraft.getInstance().gameDirectory.getAbsolutePath(), String.format("mods%snicephore", File.separator)),
            REFERENCES_JSON = new File(DESTINATION, String.format("%sreferences.json", File.separator));
    public InitThread(boolean optimiseConfig) {
        this.optimiseConfig = optimiseConfig;
    }

    @Override
    public void run() {
        if (optimiseConfig) {
            if (Files.exists(DESTINATION.toPath())) {
                {
                    try {
                        Optional<Response> response = getResponse(getJsonReader(REFERENCES_JSON));

                        if (response.isPresent()) {
                            Reference.Command.OXIPNG = response.get().oxipng_command;
                            Reference.Command.ECT = response.get().ect_command;

                            Reference.File.OXIPNG = response.get().oxipng_file;
                            Reference.File.ECT = response.get().ect_file;

                            Reference.Version.OXIPNG = response.get().oxipng_version;
                            Reference.Version.ECT = response.get().ect_version;
                        }
                    } catch (final IOException e) {
                        e.printStackTrace();
                    }
                }

                {
                    try {
                        final Optional<Response> response = getResponse(getJsonReader(Reference.DOWNLOADS_URLS, REFERENCES_JSON));

                        if (response.isPresent()) {
                            if (!Reference.Version.OXIPNG.equals(response.get().oxipng_version)) {
                                Reference.Version.OXIPNG = response.get().oxipng_version;
                                downloadAndExtract(response.get().oxipng);
                            }

                            if (!Reference.Version.ECT.equals(response.get().ect_version)) {
                                Reference.Version.ECT = response.get().ect_version;
                                downloadAndExtract(response.get().ect);
                            }
                        }
                    } catch (final IOException e) {
                        e.printStackTrace();
                    }
                }
            }
            else{
                freshInstall();
            }
        }
    }

    private static void downloadAndExtract(String urlString){

        URL url = null;
        String filename = "";

        try {
            url = new URL(urlString);
            filename = Paths.get(url.getPath()).getFileName().toString();
        } catch (IOException e) {
            e.printStackTrace();
        }

        final File zip = new File(DESTINATION, String.format("%s%s", File.separator, filename));

        if (!filename.isEmpty()) {
            try (BufferedInputStream in = new BufferedInputStream(url.openStream()); FileOutputStream fileOutputStream = new FileOutputStream(zip)) {
                byte dataBuffer[] = new byte[1024];
                int bytesRead;
                while ((bytesRead = in.read(dataBuffer, 0, 1024)) != -1) {
                    fileOutputStream.write(dataBuffer, 0, bytesRead);
                }

                Archiver archiver = ArchiverFactory.createArchiver(zip);
                archiver.extract(zip, DESTINATION);

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private Optional<Response> getResponse(final JsonReader reader) {
        final Gson gson = new Gson();
        final Type collectionType = new TypeToken<Collection<Response>>() {}.getType();
        final Collection<Response> responses = gson.fromJson(reader, collectionType);
        final Optional<Response> response = responses.stream().filter(response1 -> response1.platform.equals(Util.getOS().name())).findFirst();
        return response;
    }

    private JsonReader getJsonReader(String URL, final File file) throws IOException {
        FileUtils.copyURLToFile(new URL(URL), file);
        return getJsonReader(file);
    }

    private JsonReader getJsonReader(final File file) throws FileNotFoundException {
        return new JsonReader(new FileReader(file));
    }

    private void freshInstall() {
        try {
            Files.createDirectory(DESTINATION.toPath());
            final Optional<Response> response = getResponse(getJsonReader(Reference.DOWNLOADS_URLS, REFERENCES_JSON));

            Reference.Command.OXIPNG = response.get().oxipng_command;
            Reference.Command.ECT = response.get().ect_command;

            Reference.File.OXIPNG = response.get().oxipng_file;
            Reference.File.ECT = response.get().ect_file;

            Reference.Version.OXIPNG = response.get().oxipng_version;
            Reference.Version.ECT = response.get().ect_version;

            downloadAndExtract(response.get().oxipng);
            downloadAndExtract(response.get().ect);
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    static class Response{
        String platform;
        String oxipng, oxipng_file, oxipng_command, oxipng_version;
        String ect, ect_file, ect_command, ect_version;
    }
}

File: src/main/kotlin/com/vandendaelen/nicephore/thread/InitThread.kt
package com.vandendaelen.nicephore.thread

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonReader
import com.vandendaelen.nicephore.enums.OperatingSystems
import com.vandendaelen.nicephore.utils.Reference
import net.minecraft.client.Minecraft
import org.apache.commons.io.FileUtils
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.FileReader
import java.io.IOException
import java.net.URI
import java.nio.file.Files
import java.util.zip.ZipInputStream

class InitThread(private val optimiseConfig: Boolean) : Thread() {

    override fun run() {
        if (optimiseConfig) {
            if (Files.exists(DESTINATION.toPath())) {
                try {
                    val response = getResponse(getJsonReader(REFERENCES_JSON))
                    if (response != null) {
                        Reference.Command.OXIPNG = response.oxipng_command
                        Reference.Command.ECT = response.ect_command
                        Reference.File.OXIPNG = response.oxipng_file
                        Reference.File.ECT = response.ect_file
                        Reference.Version.OXIPNG = response.oxipng_version
                        Reference.Version.ECT = response.ect_version
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }

                try {
                    val response = getResponse(getJsonReader(Reference.DOWNLOADS_URLS, REFERENCES_JSON))
                    if (response != null) {
                        if (Reference.Version.OXIPNG != response.oxipng_version) {
                            Reference.Version.OXIPNG = response.oxipng_version
                            downloadAndExtract(response.oxipng, OXIPNG_ZIP)
                        }
                        if (Reference.Version.ECT != response.ect_version) {
                            Reference.Version.ECT = response.ect_version
                            downloadAndExtract(response.ect, ECT_ZIP)
                        }
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            } else {
                freshInstall()
            }
        }
    }

    private fun getResponse(reader: JsonReader): Response? {
        val gson = Gson()
        val collectionType = object : TypeToken<Collection<Response>>() {}.type
        val responses: Collection<Response> = gson.fromJson(reader, collectionType)
        return responses.firstOrNull { it.platform == OperatingSystems.getOS().name }
    }

    internal class Response {
        var platform: String = ""
        var oxipng: String = ""
        var oxipng_file: String = ""
        var oxipng_command: String = ""
        var oxipng_version: String = ""
        var ect: String = ""
        var ect_file: String = ""
        var ect_command: String = ""
        var ect_version: String = ""
    }

    @Throws(IOException::class)
    private fun getJsonReader(url: String, file: File): JsonReader {
        FileUtils.copyURLToFile(URI(url).toURL(), file)
        return getJsonReader(file)
    }

    @Throws(FileNotFoundException::class)
    private fun getJsonReader(file: File): JsonReader {
        return JsonReader(FileReader(file))
    }

    private fun freshInstall() {
        try {
            Files.createDirectory(DESTINATION.toPath())
            val response = getResponse(getJsonReader(Reference.DOWNLOADS_URLS, REFERENCES_JSON))

            if (response != null) {
                Reference.Command.OXIPNG = response.oxipng_command
                Reference.Command.ECT = response.ect_command
                Reference.File.OXIPNG = response.oxipng_file
                Reference.File.ECT = response.ect_file
                Reference.Version.OXIPNG = response.oxipng_version
                Reference.Version.ECT = response.ect_version

                downloadAndExtract(response.oxipng, OXIPNG_ZIP)
                downloadAndExtract(response.ect, ECT_ZIP)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        private val DESTINATION = File(
            Minecraft.getInstance().gameDirectory.absolutePath,
            "mods${File.separator}nicephore"
        )
        private val REFERENCES_JSON = File(DESTINATION, "${File.separator}references.json")
        private val OXIPNG_ZIP = File(DESTINATION, "${File.separator}oxipng.zip")
        private val ECT_ZIP = File(DESTINATION, "${File.separator}ect.zip")

        private fun downloadAndExtract(url: String, zip: File) {
            try {
                BufferedInputStream(URI(url).toURL().openStream()).use { input ->
                    FileOutputStream(zip).use { output ->
                        val dataBuffer = ByteArray(1024)
                        var bytesRead: Int
                        while (input.read(dataBuffer, 0, 1024).also { bytesRead = it } != -1) {
                            output.write(dataBuffer, 0, bytesRead)
                        }
                    }
                }
                unzip(zip.absolutePath, DESTINATION.absolutePath)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        private fun unzip(zipFilePath: String, destDir: String) {
            val dir = File(destDir)
            if (!dir.exists()) dir.mkdirs()
            val buffer = ByteArray(1024)
            try {
                FileInputStream(zipFilePath).use { fis ->
                    ZipInputStream(fis).use { zis ->
                        var ze = zis.nextEntry
                        while (ze != null) {
                            val fileName = ze.name
                            val newFile = File(destDir + File.separator + fileName)
                            println("Unzipping to ${newFile.absolutePath}")
                            File(newFile.parent).mkdirs()
                            FileOutputStream(newFile).use { fos ->
                                var len: Int
                                while (zis.read(buffer).also { len = it } > 0) {
                                    fos.write(buffer, 0, len)
                                }
                            }
                            zis.closeEntry()
                            ze = zis.nextEntry
                        }
                        zis.closeEntry()
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }
}
